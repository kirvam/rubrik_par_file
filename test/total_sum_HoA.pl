use strict;
use warnings;
use Data::Dumper;

my $pattern = "\*tv_cp_report__.csv";
my $pattern = "\*tv_cp_report__.txt";

my $event_number = 0;  
my %data;
my %table;

# my $ref = \%data;

print "--------------------< start >----------------------\n";
my(@files) = get_file_list($pattern);

my $count = $#files + 1;
print "List of Files in \@files is: ", $count ,"\n";
foreach  my $ii ( 0 .. $#files ){
  chomp($files[$ii]);
  print "[$ii] $files[$ii]\n";
}
print "\n";

my $tagf = 0;
my $valf = 3;

my $href = open_array_of_files($tagf,$valf,@files);
#print Dumper \%table;

print "Print href\n";
print Dumper $href;

my $periods = $#files;
my($legendref,$gdataref) = read_thru_HoA($periods,$href);

my @legend = @$legendref;
my @gdata = @$gdataref;

print "Legend: \@legend\n";
print Dumper \@legend;

print "GDATA: \n";
print Dumper \@gdata;


graph_it(\@legend,\@gdata);



print "---------------------< end >------------------------\n";

# SUBS
#
sub read_thru_HoA{
# headings = number of events or files.  If there is no event for data there still needs to be value like 0_0 for event 1.
# data is table which needs to have 0 valuse inserted if there is no event so it can be graphed. 
my($periods,$href) = @_;
print "Number of periods: $periods\n";
my @periods;
foreach my $ii ( 0 .. $periods ){
       push @periods, $ii;
};
print Dumper \@periods;
my @legend = ();
my @gdata = ();;
#my @garray;
print "Printing \$ref_tab!!!\n";
foreach my $key ( sort keys %{$href} ){
     print "\$key: $key: \n";
        push @legend, $key;
       my @array =  @{ ${$href}{$key} };
        my @garray;
         foreach my $ii ( 0 .. $#array ){
                     print "$array[$ii]\n";
              push @garray, $array[$ii];       
                 }
           push @gdata, [ @garray ];
      #print "\$table{\$key}: $table{$key}\n";
 };
  unshift @gdata, [ @periods ];
  return(\@legend,\@gdata);
}

sub open_array_of_files{
my($tagf,$valf,@list) = @_;
my %data;
  print "\$tagf=$tagf\n\$valf=$valf\n";
    my $tag = ();
    my $val = (); 
 foreach my $ii ( 0 .. $#list ){
    my $file = $list[$ii];
    print "opening: $file\n";
  open ( my $fh, '<' , $file) || die "Flaming death on open of $file: $?\n";
    while(<$fh>){
    my $line = $_;
    chomp($line);
     print "\$line: $line\n";
    if( $line =~ m/(heading)/ig ) {
       next;
       print $1,"\n";
      }
    my @array = split(/,/,$line);
     $tag = $array[$tagf];
     $val = $array[$valf];
      my $entry = $array[$valf];
#     $data{$tag} += $array[$valf];
    $data{$tag}[$ii] = $entry;
###  push( @{ $data{$tag} }, "$ii"."_".$val );
   };
       print Dumper \%data;
    #foreach my $tag ( sort keys %data ){
    #  print "\$key: $tag\n";
    #   my $value = $data{$tag};
    #   $value = $ii."_".$value;
    #  push( @{ $table{$tag} }, $value );  
    #}; 
 } return(\%data);
};

sub get_file_list{
my($pattern) = @_;
my $cmd = "/bin/ls ";
$cmd = $cmd."$pattern";
my @files = `$cmd`;
#print "List of report files:\n";
my @sorted = sort @files;
foreach my $item (@sorted){
  chomp($item);
 # print $item,"\n";
  }
return(@sorted);
};
 

sub get_date{
my ($sec,$min,$hour,$mday,$mon,$year,$wday,$yday,$isdst) =
                                                localtime(time);
$year += 1900;
#print $year,"\n";
$mon += 1;
#$year = sprintf("%02d", $year % 100);
$mday = sprintf("%02d", $mday);
my $date = $year."-".$mon."-".$mday;
return ($date);
};


sub sum_sub {
my($ref) = @_;
my $ct =();
foreach my $key ( keys %{ $ref } ) {
       print "\$key: ";
       print $key,"| ";
     foreach my $ii ( 0 .. $#{ @{$ref}{$key} } ){
      my @array = @{  ${$ref}{$key}[$ii] };
       foreach my $jj ( 0 .. $#array ){
        #print "  ";
        #print "[$ii][$jj]: ",$array[$jj],"\n";
        # grab the 3rd value
        if( $jj eq 3 ){
        $ct += $array[$jj];
        }
    }
   }
       push( @{ $data{$key} }, "$event_number"."_".$ct );
       #$data{$key} = @[ $ct ];
       print "sum: $ct\n";
 };
       $event_number++;


};

sub graph_it{
my($legend,$data) = @_;
print "Dumping \$legend ref\n";
print Dumper \$legend;

print "Dumping \$data ref\n";
print Dumper \$data;

use GD::Graph::linespoints;
use GD::Graph::lines;
use GD::Graph::hbars;

my $graph = GD::Graph::linespoints->new(700, 600);

$graph->set(
    x_label           => 'Samples taken over Time',
    y_label           => 'GB\s Capacity Used on the Rubriks',
    title             => 'Rubrik Capacity Graph',
    y_max_value       => 200,
    y_tick_number     => 50,
    y_label_skip      => 5,

) or die $graph->error;

$graph->set_legend_font(GD::Font->Tiny);
$graph->set_legend(@$legend);

my $gd = $graph->plot($data) or die $graph->error;

open(IMG, '>RBK1.png') or die $!;
binmode IMG;
print IMG $gd->png;
close IMG;

open(IMG, '>RBK1.gif') or die $!;
binmode IMG;
print IMG $gd->gif;
close IMG;

print "finished with graph creation!!!\n";
};
