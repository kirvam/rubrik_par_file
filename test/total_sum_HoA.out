--------------------< start >----------------------
List of Files in @files is: 5
[0] 2018-10-29__tv_cp_report__.txt
[1] 2018-11-02__tv_cp_report__.txt
[2] 2018-11-05__tv_cp_report__.txt
[3] 2018-11-13__tv_cp_report__.txt
[4] 2018-11-15__tv_cp_report__.txt

$tagf=0
$valf=3
opening: 2018-10-29__tv_cp_report__.txt
$line: # Heading
$line: AAA,bulldog,chow,15
$line: AAA,elkhound,chow,13
$line: AAA,shepard,dog-chow,23
$line: ABC,bull,chow,11
$line: ABC,stafford,oats,18
$line: ABC,black,wheat,50
$line: XXX,tiger,deer,30
$line: XXX,lion,gazelle,33
$line: XXX,puma,mule-deer,45
$line: XXX,hobbes,tunafish,55
$VAR1 = {
          'XXX' => [
                     '55'
                   ],
          'ABC' => [
                     '50'
                   ],
          'AAA' => [
                     '23'
                   ]
        };
opening: 2018-11-02__tv_cp_report__.txt
$line: # Heading
$line: AAA,bulldog,chow,25
$line: AAA,elkhound,chow,15
$line: AAA,shepard,dog-chow,27
$line: ABC,bull,chow,13
$line: ABC,stafford,oats,19
$line: ABC,black,wheat,55
$line: XXX,tiger,deer,34
$line: XXX,lion,gazelle,36
$line: XXX,puma,mule-deer,40
$line: XXX,calvin,sugar-shockers,51
$line: ZZZ,hobbes,tunafish,67
$VAR1 = {
          'XXX' => [
                     '55',
                     '51'
                   ],
          'ABC' => [
                     '50',
                     '55'
                   ],
          'AAA' => [
                     '23',
                     '27'
                   ],
          'ZZZ' => [
                     undef,
                     '67'
                   ]
        };
opening: 2018-11-05__tv_cp_report__.txt
$line: # Heading
$line: AAA,bulldog,chow,27
$line: AAA,elkhound,chow,14
$line: AAA,shepard,dog-chow,26
$line: ABC,stafford,oats,29
$line: ABC,black,wheat,57
$line: XXX,tiger,deer,39
$line: XXX,lion,gazelle,38
$line: XXX,puma,mule-deer,43
$line: XXX,calvin,sugar-shockers,53
$line: ZZZ,hobbes,tunafish,68
$VAR1 = {
          'XXX' => [
                     '55',
                     '51',
                     '53'
                   ],
          'ABC' => [
                     '50',
                     '55',
                     '57'
                   ],
          'AAA' => [
                     '23',
                     '27',
                     '26'
                   ],
          'ZZZ' => [
                     undef,
                     '67',
                     '68'
                   ]
        };
opening: 2018-11-13__tv_cp_report__.txt
$line: # Heading
$line: AAA,bulldog,chow,27
$line: AAA,elkhound,chow,14
$line: AAA,shepard,dog-chow,26
$line: XXX,tiger,deer,39
$line: XXX,puma,mule-deer,43
$line: XXX,calvin,sugar-shockers,53
$line: ZZZ,hobbes,tunafish,68
$VAR1 = {
          'XXX' => [
                     '55',
                     '51',
                     '53',
                     '53'
                   ],
          'ABC' => [
                     '50',
                     '55',
                     '57'
                   ],
          'AAA' => [
                     '23',
                     '27',
                     '26',
                     '26'
                   ],
          'ZZZ' => [
                     undef,
                     '67',
                     '68',
                     '68'
                   ]
        };
opening: 2018-11-15__tv_cp_report__.txt
$line: # Heading
$line: AAA,bulldog,chow,27
$line: AAA,elkhound,chow,14
$line: AAA,shepard,dog-chow,26
$line: ABC,stafford,oats,29
$line: ABC,black,wheat,57
$line: XXX,tiger,deer,39
$line: XXX,lion,gazelle,38
$line: XXX,puma,mule-deer,43
$line: XXX,calvin,sugar-shockers,53
$line: ZZZ,hobbes,tunafish,68
$VAR1 = {
          'XXX' => [
                     '55',
                     '51',
                     '53',
                     '53',
                     '53'
                   ],
          'ABC' => [
                     '50',
                     '55',
                     '57',
                     undef,
                     '57'
                   ],
          'AAA' => [
                     '23',
                     '27',
                     '26',
                     '26',
                     '26'
                   ],
          'ZZZ' => [
                     undef,
                     '67',
                     '68',
                     '68',
                     '68'
                   ]
        };
Print href
$VAR1 = {
          'XXX' => [
                     '55',
                     '51',
                     '53',
                     '53',
                     '53'
                   ],
          'ABC' => [
                     '50',
                     '55',
                     '57',
                     undef,
                     '57'
                   ],
          'AAA' => [
                     '23',
                     '27',
                     '26',
                     '26',
                     '26'
                   ],
          'ZZZ' => [
                     undef,
                     '67',
                     '68',
                     '68',
                     '68'
                   ]
        };
Number of periods: 4
$VAR1 = [
          0,
          1,
          2,
          3,
          4
        ];
Printing $ref_tab!!!
$key: AAA: 
23
27
26
26
26
$key: ABC: 
50
55
57

57
$key: XXX: 
55
51
53
53
53
$key: ZZZ: 

67
68
68
68
Legend: @legend
$VAR1 = [
          'AAA',
          'ABC',
          'XXX',
          'ZZZ'
        ];
GDATA: 
$VAR1 = [
          [
            0,
            1,
            2,
            3,
            4
          ],
          [
            '23',
            '27',
            '26',
            '26',
            '26'
          ],
          [
            '50',
            '55',
            '57',
            undef,
            '57'
          ],
          [
            '55',
            '51',
            '53',
            '53',
            '53'
          ],
          [
            undef,
            '67',
            '68',
            '68',
            '68'
          ]
        ];
Dumping $legend ref
$VAR1 = \[
            'AAA',
            'ABC',
            'XXX',
            'ZZZ'
          ];
Dumping $data ref
$VAR1 = \[
            [
              0,
              1,
              2,
              3,
              4
            ],
            [
              '23',
              '27',
              '26',
              '26',
              '26'
            ],
            [
              '50',
              '55',
              '57',
              undef,
              '57'
            ],
            [
              '55',
              '51',
              '53',
              '53',
              '53'
            ],
            [
              undef,
              '67',
              '68',
              '68',
              '68'
            ]
          ];
finished with graph creation!!!
---------------------< end >------------------------
